plugins/usbdmx/AsyncPluginImpl.cpp

```diff
@@ -353,17 +346,26 @@ bool AsyncPluginImpl::StartAndRegisterDevice(Widget *widget, Device *device) {
 
 /*
  * @brief Signal widget removal.
- * @param device_id The device id to remove.
+ * @param state The DeviceState owning the widget to be removed.
  *
- * This is run within the main thread.
+ * This must be run within the main thread. The Device destructor below may
+ * cause libusb_close() to be called, which would deadlock if the hotplug
+ * event thread were to wait for it, say in AsyncPluginImpl::DeviceEvent().
  */
-void AsyncPluginImpl::ShutdownDevice(Device *device, Future<void> *f) {
-  m_plugin_adaptor->UnregisterDevice(device);
-  device->Stop();
-  delete device;
-  if (f) {
-    f->Set();
+void AsyncPluginImpl::ShutdownDeviceState(DeviceState *state) {
+  if (state->ola_device) {
+    Device *device = state->ola_device;
+    m_plugin_adaptor->UnregisterDevice(device);
+    device->Stop();
+    delete device;
+    state->ola_device = NULL;
+  } else {
+          /* This case can be legitimate when the widget setup through the
+           * widget factory is delayed and an unplug event happens before
+           * that is completed. */
+          OLA_DEBUG << "ola_device was NULL at shutdown";
```

Comment:
Because of the push I can't tell if I missed this before or it's newly introduced:
```suggestion
    /* This case can be legitimate when the widget setup through the
     * widget factory is delayed and an unplug event happens before
     * that is completed. */
    OLA_DEBUG << "ola_device was NULL at shutdown";
```

Reply:
No that was likely alread there. Done.